### PR TITLE
docs - update data sandboxing

### PR DESCRIPTION
**Pull request by nllho**
Overall goal is to make the doc more modular, so that:
- People can see how sandboxes interact with groups, collections, other permissions, etc.
- It will be easier (down the line) to write about complex data sandboxing setups.

To do:
- [x] Copy the changes to the actual `data-sandboxes.md`
- [ ] Add links from other docs to new sandboxing doc sections (will do in separate PR)
- [ ] Rename "advanced sandboxing" in Learn (if needed)

---

The plan (and ongoing research):
https://www.notion.so/metabase/Data-sandboxing-parent-ticket-ed6e50ca364a45aebc3d46d81195a9e9

This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30445)

### CODE REVIEW

**Review by jeff-bruemmer — COMMENTED** [review]
Left a lot of comments, but nice work so far. These docs needed an update, and I like how you're breaking down how sandboxes interact with groups, collections, other permissions, etc.

> Any objections to the "row-limiting sandbox" (renaming the "regular" sandbox)? I think it is easier to think about permissions when each permission is named after the thing that is being giveth or taketh away. "Removing a row limit" vs. "removing the column used to filter the table"?

No objections. Same with column-limiting; see my notes. Though we may want to use “restrict”, since ‘LIMIT‘ is a thing in SQL.

File: docs/permissions/data-sandboxes-draft.md [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+### Row-limiting data sandboxes
+
+A row-limiting sandbox displays the result of a filtered table. Groups assigned to a row-limiting sandbox will see a version of the table with some `Filter = Value` applied, instead of the original table.
+
+For example, you can create a row-limiting sandbox to filter the Accounts table so that one person sees the rows where `Plan = Basic`, and another person only sees the rows where `Plan = Premium`. 
+
+The data sandbox will dynamically set the filter value to "Basic" or "Premium" to hide different rows for each person based on their [user attributes](../people-and-groups/managing.md#adding-user-attributes).
+
+### Advanced data sandboxes
```
[the remark this answers is not in the document]
Comment (nllho) [reply docs/permissions/data-sandboxes-draft.md:35]: I thought about this more. I think that the "advanced sandbox" is more like a "custom sandbox" because you can put anything in the saved SQL question.

Restricting columns will be the most common use case, but you can do a lot of other useful security things. Like displaying a hashed column instead of an original column. Or aggregating a survey results table and displaying responses where the count is > 5 (common when dealing with PII).

Basically I'm thinking of putting this line into the planned Learn tutorials:
> [Create a custom sandbox]() for Group A so that they can only see columns X and rows Y.

File: docs/permissions/data-sandboxes-draft.md [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+To display different filtered versions of the saved question for each person in an advanced data sandbox, set up [user attributes](../people-and-groups/managing.md#adding-user-attributes) as well.
+
+## Creating an advanced sandbox
+
+> Make sure to do the [prerequisites](#prerequisites-for-advanced-data-sandboxes) first.
+
+1. Go to **Admin settings** > **Permissions**.
+2. Select the database and table that you want to sandbox.
+3. Find the group that you want to put in the sandbox.
+4. Disable **Native query editing** for the group.
```
[the remark this answers is not in the document]
Comment (nllho) [reply docs/permissions/data-sandboxes-draft.md:88]: Good catch! I was working out how to make sure people catch all the prerequisites, but you're right that Native Querying will automatically get disabled.

File: docs/permissions/data-sandboxes-draft.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+You can define one data sandbox (that is, one permission set) for each table and group pairing in your Metabase.
+
+\* SQL questions cannot be sandboxed. See [Preventing data sandboxing permissions conflicts](#preventing-data-sandboxing-permissions-conflicts).
+
+## Types of data sandboxes
+
+For each group and table pairing, you can choose between:
+
+- A [row-limiting data sandbox](#row-limiting-data-sandboxes), to hide rows from a table based on a user attribute.
+- An [advanced data sandbox](#advanced-data-sandboxes), to display a custom query result in place of a table (the attribute-based filter is optional).
```
[the remark this answers is not in the document]
Comment (nllho) [reply docs/permissions/data-sandboxes-draft.md:25]: For the types overview, I think we should focus on the outcome over the mechanism.

The thing we need to solve (aside from smoothing out setup mishaps) is that lots of people actually misunderstand what a sandbox IS.

So I'm trying to dial in on a very minimal description, like "a sandbox lets you replace A with B for a group of people". I think the "find and replace" idea will help people visualize how a sandbox works more fundamentally (before getting into filters, attributes, etc).

I also think that our description of the row-restricting sandbox (filtering using a column) seems to cause the common misunderstanding that it's the "columns that are filtered", so I'm deliberately trying to keep those words a little farther away from the top-level descriptions 😅

**Comment by nllho** [comment]
@jeff-bruemmer Typing this out for posterity :) 

For more context about the suggested structure/naming in this PR:

The goal is to set up reusable doc "tasks" that make sense in a [variety of use cases](https://github.com/metabase/metabase.github.io/pull/2561).

I think we should make the data sandboxing tasks _just_ abstract enough so that they make sense when placed in different concrete examples in Learn.

To help people remember what a sandbox DOES (among all the complexity of setting one up), I broke things up like this:
- For some group, find table A, and replace it with table B. 
- Configure table B using filters, attributes, etc.

I think we should keep the name "advanced sandbox" for now -- it is a bit misleading to call it a "column sandbox". Agree that the row vs. column deciding factor should be clear up front (everywhere else besides the name!).

File: docs/permissions/data-sandboxes-draft.md [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+### Displaying edited columns in an advanced sandbox
+
+Aside from excluding rows and columns from an advanced sandbox, you can also **display edited columns** (without changing the columns in your database). 
+
+For example, you can create a "Sandboxed Accounts" SQL question that truncates the Email column to display usernames instead of complete email addresses.
+
+You cannot add columns to an advanced sandbox. The schema of the saved SQL question (the question you want to display in the sandbox) must match the schema of the original table.
+
+In our example, that means the "Sandboxed Accounts" SQL question must return the same number of columns and corresponding data types as the original Accounts table.
```
Comment (jeff-bruemmer) [inline docs/permissions/data-sandboxes-draft.md:129]: Or fewer columns?

File: docs/permissions/data-sandboxes-draft.md [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+## Prerequisites for advanced sandboxes
+
+- A [group](../people-and-groups/managing.md#groups) of people to be added to the advanced data sandbox.
+- An admin-only [collection](../exploration-and-organization/collections.md) ([collection permissions](../permissions/collections.md) set to **No access** for all groups except Administrators).
+- A [saved SQL question](../people-and-groups/) with the rows and columns to be displayed to the people in the advanced sandbox, stored in the admin-only collection.
+- Optional: [User attributes](../people-and-groups/managing.md#adding-a-user-attribute) for each of the people in the group.
+
+To display different filtered versions of the saved question for each person in an advanced data sandbox, set up [user attributes](../people-and-groups/managing.md#adding-user-attributes) as well.
+
+### Designing a saved question to be displayed in an advanced sandbox
```
Comment (jeff-bruemmer) [inline docs/permissions/data-sandboxes-draft.md:113]:
```suggestion
### Create a SQL question that Metabase can use to create a custom view of the table you want to sandbox
```

File: docs/permissions/data-sandboxes-draft.md [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+4. Find the group that you want to put in the sandbox.
+5. Click on the dropdown under **Data access** for that group.
+6. Select "Sandboxed".
+7. Select "Use a saved question to create a custom view for this table".
+8. Select your saved question.
+9. Optional: [restrict rows based on people's user attributes](#restricting-rows-in-an-advanced-sandbox-with-user-attributes) to the saved question.
+
+> If you have saved SQL questions that use sandboxed data, make sure to move all of those questions to admin-only collections. For more info, see [Permissions conflicts: saved SQL questions](#saved-sql-questions).
+
+For an example, see [Advanced data sandboxing: limiting access to columns](https://www.metabase.com/learn/permissions/data-sandboxing-column-permissions).
```
Comment (jeff-bruemmer) [inline docs/permissions/data-sandboxes-draft.md:145]: Just noting that if we ditch "Advanced", we'll need to update the learn articles as well.

File: docs/permissions/data-sandboxes-draft.md [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+```
+WHERE plan = USER_ATTRIBUTE_VALUE
+```
+
+Metabase replaces the SQL parameter with the specific **user attribute value** (such as "Basic") associated with a sandboxed person. When that person logs into Metabase and uses the sandboxed table, they'll see the query result that uses:
+```
+WHERE plan = "Basic"
+```
+
+## How sandboxing permissions interact with other permissions
```
Comment (jeff-bruemmer) [inline docs/permissions/data-sandboxes-draft.md:192]: Now I'm thinking we're oversharing here, as in giving people way too much to read that they should haven't to.

The important info here is where sandboxing rules break down, or get iffy.

Everything you have in with "uses sandboxing rules" set to "always" we can just gloss over. In fact, we shouldn't mention it. We can say something like "Sandboxing rules apply everywhere in Metabase for a group, with the following caveats:" And then just stick to the edge cases that people might need to consider. We shouldn't, for example, make them read a section on the Data Browser to figure out that, yes, looks like we're all good there.

File: docs/permissions/data-sandboxes-draft.md [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+title: Data sandboxes
+---
+
+# Data sandboxes
+
+{% include plans-blockquote.html feature="Data sandboxes" %}
+
+Data sandboxes let you give granular permissions to rows and columns for different groups of people.
+
+Say you have people who you want to be able to log into your Metabase, but who should only be able to view data that pertains to them. For example, you might want to display an Accounts table to some customers, but each customer should only be able to view the rows that match their customer ID. Sandboxes let you give people a way to self-serve their own analytics, without ever seeing any results that don't pertain to them.
```
Comment (jeff-bruemmer) [inline docs/permissions/data-sandboxes-draft.md:11]: Maybe link to https://www.metabase.com/learn/customer-facing-analytics/multi-tenant-self-service-analytics for context.

**Review by jeff-bruemmer — CHANGES_REQUESTED** [review]
Left some comments regarding that final section, and I'm still not sold on the naming here, but nice work here, @nllho.

File: docs/permissions/data-sandboxes-draft.md (hunk 5 above)
Comment (nllho) [reply docs/permissions/data-sandboxes-draft.md:113]: I have been writing "displays X" instead of "view of X" to avoid confusing a sandboxed table with a materialized view of that table.

Though I wasn't sure if Metabase _was_ creating a materialized view for a sandboxed table. But I just looked it up, and views require CREATE privileges. Luckily, we worked on privileges recently, so I'm pretty sure the only feature that requires CREATE is model caching.

File: docs/permissions/data-sandboxes-draft.md (hunk 7 above)
Comment (nllho) [reply docs/permissions/data-sandboxes-draft.md:192]: Good point! I wrote it out like this because if I was the person configuring permissions for sensitive data, I would want to check each place where the sandboxed table appears (just to make sure I did the permissions correctly).

How about we keep a comprehensive list in the table (so people can verify if they want), but eliminating the detailed sections for the "Always" line items?

File: docs/permissions/data-sandboxes-draft.md [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+
+A group assigned to a row-restricted sandbox will see a table with some `Filter = "User Attribute"` applied.
+
+For example, you can create a row-restricted sandbox to filter an Accounts table for a sandboxed group so that:
+- A sandboxed person with the user attribute "Basic" will see the rows where `Plan = "Basic"`.
+- A sandboxed person with the user attribute "Premium" will only see the rows where `Plan = "Premium"`.
+
+For more info, see [Choosing user attributes for sandboxing](#choosing-user-attributes-for-sandboxing).
+
+### Advanced sandboxes
```
[the remark this answers is not in the document]
Comment (nllho) [reply docs/permissions/data-sandboxes-draft.md:42]: What about "custom sandbox"?

I'm still trying to avoid boxing in (ha) the second sandbox type. 

It isn't quite a "column sandbox" (because you can restrict rows), but it isn't strictly "row and column-restricting sandbox" either (because you can edit a column instead of excluding it). 

You can also create an advanced sandbox JUST for the sake of passing user attributes to parameters in Metabase. (I'm thinking of the [Markdown URL](https://www.metabase.com/learn/dashboards/markdown#create-a-custom-url-with-a-filter-value) use case). 

Of course, it is very useful for people to access user attributes from parameters without necessarily enforcing permissions. But you cannot currently pass a user attribute directly to a parameter in Metabase -- you have to set up a sandbox in the middle. (A success eng candidate actually pointed that out to us recently). In that case, it seems easier to tell people to create a "custom sandbox" that doesn't apply any permissions -- just `SELECT *` from table.

All of this to say that I don't know if parameter functionality will evolve faster than sandboxes (unless we have a read on that from product?). So I think we may want to keep the name future-proof for just a bit longer.

[the remark this answers is not in the document]
Comment (jeff-bruemmer) [reply docs/permissions/data-sandboxes-draft.md]: You're right. The "handle" is important. And I like "custom sandbox", and think it's sufficiently different to merit the change (especially given we only use "Advanced" in the docs/learn). Advanced implies more capable, but can also suggest a higher degree of difficulty; "custom" implies flexibility, tailoring, etc, without implication of requiring a lot of knowledge or expertise.

So how about something like:

```
### Basic data sandboxes (or row-restricted sandboxes, or row-based data sandboxes).

These sandboxes filter by a column in the table based on user attributes.

Use a [NAME] sandbox when you need to:

- Restrict rows based on user attributes
  
### Custom data sandboxes

These sandboxes uses a saved question to create a custom view for tables.

Use a custom data sandbox when you need to:

- Restrict rows based on user attribute
- Restrict columns based on user attributes
- Create custom column views (like only show the username of an email)
- Pass user attributes as parameters... (link to relevant docs)
```

Or whatever. Given our conversation, the constraints for this quick overview seem like they should be:

- List use cases so people know which box to choose.
- Give people a handle to talk about each method.
- Reflect language used in Metabase itself.
- Summarize how they're setup/how they work.

So, rewrite with those constraints in mind.

> You cannot currently pass a user attribute directly to a parameter in Metabase -- you have to set up a sandbox in the middle. (A success eng candidate actually pointed that out to us recently). In that case, it seems easier to tell people to create a "custom sandbox" that doesn't apply any permissions -- just SELECT * from table.

Is the basic, filter-by-column sandbox insufficient for this? Or is the idea that you want to be able to pass parameters _without_ restricting views?

File: docs/permissions/data-sandboxes-draft.md [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+### Multiple data sandboxing permissions
+
+Multiple data sandboxes on the same table can create a permissions conflict. You can add a person to a maximum of one data sandbox per table (via the person's group membership).
+
+For example, if you have:
+
+- One data sandbox for the group "Basic Accounts" that filters the Accounts table on `Plan = "Basic"`.
+- Another data sandbox for the group "Converted Accounts" that filters the Accounts table on `Trial Converted = true`.
+
+If you put Vincent Accountman in both groups, he'll be in conflicting sandboxes for the Accounts table, and get an error message whenever he tries to use Accounts in Metabase.
```
Comment (jeff-bruemmer) [inline docs/permissions/data-sandboxes-draft.md:230]: > Vincent Accountman

File: docs/permissions/data-sandboxes-draft.md [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+| Data reference                     | Always                                                          |
+| Public links                       | [Depends on other permissions](#public-sharing)                 |
+| Query builder                      | Always                                                          |
+| Saved question (GUI)               | Always                                                          |
+| Saved question (SQL)               | [Depends on other permissions](#saved-sql-questions)            |
+| Subscriptions or alerts (sending)  | [Depends on other permissions](../permissions/notifications.md) |
+| Subscriptions or alerts (receiving | Always                                                          |
+| SQL editor                         | [Never](#limitations)                                           |
+
+## Preventing data sandboxing permissions conflicts
```
Comment (jeff-bruemmer) [inline docs/permissions/data-sandboxes-draft.md:211]: This section is much better. Nice work.

File: docs/permissions/data-sandboxes-draft.md [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+```
+WHERE plan = USER_ATTRIBUTE_VALUE
+```
+
+Metabase replaces the SQL parameter with the specific **user attribute value** (such as "Basic") associated with a sandboxed person. When that person logs into Metabase and uses the sandboxed table, they'll see the query result that uses:
+```
+WHERE plan = "Basic"
+```
+
+## How sandboxing permissions interact with other permissions
```
Comment (jeff-bruemmer) [inline docs/permissions/data-sandboxes-draft.md:195]: I'm still not sure we need this section. Rather: it's better to just list the scenarios, like you do below. By including this section (its table in particular), we needlessly risk incompletion. What about models? actions? etc.

If we _only_ list the scenarios where things will not respect the sandboxes, we should be good, and it's less info for people to have to read.

**Review by jeff-bruemmer — APPROVED** [review]
Great revision, @nllho. Approving to unblock you, but let me know if you have any questions [re: naming](https://github.com/metabase/metabase/pull/30445#discussion_r1189837352).

File: docs/permissions/data-sandboxes-draft.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (nllho) [reply docs/permissions/data-sandboxes-draft.md]: > Is the basic, filter-by-column sandbox insufficient for this? Or is the idea that you want to be able to pass parameters without restricting views?

Yeah, the second thing! 

You can imagine -- it is super useful to be able to access a user attribute from a parameter. For now, it would be great to show people that they can link those things together (via a sandbox), even if they have no need for the permissions part.

File: docs/permissions/data-sandboxes-draft.md (hunk 12 above)
Comment (nllho) [reply docs/permissions/data-sandboxes-draft.md:195]: Yeah, I let this sink in some more. I think I was hanging onto this approach because I spend so much time in troubleshooting-land, where going through each case is very useful! 

Agree that it is unnecessary for the docs :)
